torchrl/modules/tensordict_module/rnn.py

```diff
@@ -399,3 +400,379 @@ def _lstm(
+
+        Examples:
+            >>> from torchrl.envs import GymEnv, TransformedEnv, InitTracker, step_mdp
+            >>> from torchrl.modules import MLP
+            >>> from tensordict import TensorDict
+            >>> from torch import nn
+            >>> from tensordict.nn import TensorDictSequential as Seq, TensorDictModule as Mod
+            >>> env = TransformedEnv(GymEnv("Pendulum-v1"), InitTracker())
+            >>> gru = nn.GRU(input_size=env.observation_spec["observation"].shape[-1], hidden_size=64, batch_first=True)
+            >>> gru_module = GRUModule(gru=gru, in_keys=["observation", "hidden"], out_keys=["intermediate", ("next", "hidden")])
+            >>> mlp = MLP(num_cells=[64], out_features=1)
+            >>> # building two policies with different behaviours:
+            >>> policy_inference = Seq(gru_module, Mod(mlp, in_keys=["intermediate"], out_keys=["action"]))
+            >>> policy_training = Seq(gru_module.set_recurrent_mode(True), Mod(mlp, in_keys=["intermediate"], out_keys=["action"]))
+            >>> traj_td = env.rollout(3) # some random temporal data
+            >>> traj_td = policy_training(traj_td)
+            >>> # let's check that both return the same results
+            >>> td_inf = TensorDict({}, traj_td.shape[:-1])
+            >>> for td in traj_td.unbind(-1):
+            ...     td_inf = td_inf.update(td.select("is_init", "observation", ("next", "observation")))
+            ...     td_inf = policy_inference(td_inf)
+            ...     td_inf = step_mdp(td_inf)
+            ...
+            >>> torch.testing.assert_close(td_inf["hidden"], traj_td[..., -1]["next", "hidden"])
+        """
+        if mode is self._temporal_mode:
+            return self
+        out = GRUModule(gru=self.gru, in_keys=self.in_keys, out_keys=self.out_keys)
+        out._temporal_mode = mode
+        return out
+
+    def forward(self, tensordict: TensorDictBase):
+        # we want to get an error if the value input is missing, but not the hidden states
+        defaults = [NO_DEFAULT, None]
+        shape = tensordict.shape
+        tensordict_shaped = tensordict
+        if self.temporal_mode:
+            # if less than 2 dims, unsqueeze
+            ndim = tensordict_shaped.get(self.in_keys[0]).ndim
+            while ndim < 3:
+                tensordict_shaped = tensordict_shaped.unsqueeze(0)
+                ndim += 1
+            if ndim > 3:
+                dims_to_flatten = ndim - 3
+                # we assume that the tensordict can be flattened like this
+                nelts = prod(tensordict_shaped.shape[: dims_to_flatten + 1])
+                tensordict_shaped = tensordict_shaped.apply(
+                    lambda value: value.flatten(0, dims_to_flatten),
+                    batch_size=[nelts, tensordict_shaped.shape[-1]],
+                )
+        else:
+            tensordict_shaped = tensordict.reshape(-1).unsqueeze(-1)
+
+        is_init = tensordict_shaped.get("is_init").squeeze(-1)
```

Comment:
here we are assuming `is_init` has the tensordict shape [*B,T]?

Reply:
not sure what makes you think that? If so is it a problem? Can you elaborate?

Reply:
- What makes me think that is that we consider its last dimension as T in the lines after
- What might become a problem is that 1) we are hardcoding the name "is_init". 2) it might have a different shape as it follows a done no?

Reply:
I think this issue and the one above boil down to what we want to accept as input to this module.
We need to have a way to tell what the time dimension is.
- One option is to check the dim names but it's hard to get people to adopt that. I tried and got multiple issues of people saying that their td was lacking the "time" dim name and they did not know what to do.
- Another option is to pass the time dim in the constructor but it's brittle. If you say `time_dim=3` you won't be able to change the batch size, if you say `time_dim=-3` you need to say with respect to what (obs, init, tensordict?)

A regular usage is assumed to be that the time is the last dim of the tensordict, but it's true that we could cover others. Otherwise we assume that the TD provided is always the root and always has time dim at the end. Handling batch-dims after the time dim can also be considered.

Another issue is that we assume that the hidden state can more or less be indexed by is_init but i'm not sure what this looks like in cases where you have multiple done states and multiple observations. The batch size of the obs could differ from the done state (eg in VMAS IIUC you have one done state per group of agents but one obs per agent). 
In MARL settings you will also need to consider if the LSTM reads the aggregated observations of not, if you have one LSMT per agent or not etc. Overall it starts to look like a different class to me.

TL;DR 
Happy to have a follow-up work on GRU and LSTM to make them compatible with more convoluted data structures, but do you think we can integrate GRU as it is now since it applies to most single agents use cases already?
If you have a precise list of queries of what this class should also cover i'd be glad to look at it. 
What I would need is: what are the data structures that need to be considered, what behaviour we expect, and what we want to explicitly forbid because there is a solution. To give you an example, IMO the example of the done that has a shape that differs from the obs can be easily solved by forbidding that and telling users to append an rename + unsqueeze + expand transform before (which all work as regular nn.Modules).

Reply:
> i would assume that is_init, the input, and the recurrent state have all the same batch_size and the last dim respectively of 1, F, H. In the example above, is_init might have to be expanded into the group before passing it to the module

my view in this case is (1) raise an exception because we can't really tell what to do anticipatively with a is_init that has a mismatching shape and (2) provide info on how to do the shape modif

Reply:
If only first batch dims were working with rnns, sigh!